graphstorm-processing/graphstorm_processing/graph_loaders/dist_heterogeneous_loader.py

```diff
@@ -0,0 +1,1560 @@
+
+    Parameters
+    ----------
+    spark : SparkSession
+        The SparkSession that we use to perform the processing
+    local_input_path : str
+        Local path to input configuration data
+    local_metadata_path : str
+        Local path to where the output metadata files will be created.
+    data_configs : Dict[str, Sequence[StructureConfig]]
+        Dictionary of node and edge configurations objects.
+    input_prefix : str
+        The prefix to the input data. Can be an S3 URI or an **absolute** local path.
+    output_prefix : str
+        The prefix to where the output data will be created. Can be an S3 URI
+        or an **absolute** local path.
+    num_output_files : Optional[int], optional
+        The number of files (partitions) to create for the output, if None we
+        let Spark decide.
+    enable_assertions : bool, optional
+        When true enables sanity checks for the output created.
+        However these are costly to compute, so we disable them by default.
+    """
+
+    def __init__(
+        self,
+        spark: SparkSession,
+        local_input_path: str,
+        local_output_path: str,
+        data_configs: Dict[str, Sequence[StructureConfig]],
+        input_prefix: str,
+        output_prefix: str,
+        num_output_files: Optional[int] = None,
+        add_reverse_edges=True,
+        enable_assertions=False,
+    ):
+        super().__init__(local_input_path, local_output_path, data_configs)
+
+        # TODO: Pass as an argument?
+        if input_prefix.startswith("s3://"):
+            self.filesystem_type = "s3"
+        else:
+            assert os.path.isabs(input_prefix), "We expect an absolute path"
+            self.filesystem_type = "local"
+
+        self.spark = spark  # type: SparkSession
+        self.add_reverse_edges = add_reverse_edges
+        # Remove trailing slash in s3 paths
+        if self.filesystem_type == "s3":
+            self.input_prefix = input_prefix[:-1] if input_prefix[-1] == "/" else input_prefix
+            self.output_prefix = output_prefix[:-1] if output_prefix[-1] == "/" else output_prefix
+        else:
+            # TODO: Any checks for local paths?
+            self.input_prefix = input_prefix
+            self.output_prefix = output_prefix
+        self.num_output_files = (
+            num_output_files
+            if num_output_files and num_output_files > 0
+            else int(spark.sparkContext.defaultParallelism)
+        )
+        # Mapping from node type to filepath, each file is a node-str to node-int-id mapping
+        self.node_mapping_paths = {}  # type: Dict[str, Sequence[str]]
+        # Mapping from label name to value counts
+        self.label_properties: Dict[str, Counter] = defaultdict(Counter)
+        self.timers = defaultdict(float)  # type: Dict
+        self.enable_assertions = enable_assertions
+        self.column_substitutions = {}  # type: Dict[str, str]
```

Comment:
It can happen that a column name that is valid in CSV is not valid in Parquet, see https://stackoverflow.com/a/53149013/209882

What we do in this dictionary is keep track of any column names that needed to be changed, and then we save this dict as a JSON file along with the rest of the output.

Reply:
Can you add some docstring to explain this in the code?